### PR TITLE
feat(text-area):add textarea component

### PR DESCRIPTION
**Pull request by crstnmac**
<image src="https://user-images.githubusercontent.com/20360870/200476857-9478521c-35f3-4bac-83b1-a7f0075842f5.png" width="400" height="900" />

### CODE REVIEW

File: example/src/modules/forms/TextAreaScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import React from "react";
```
Comment (Karthik-B-06) [inline example/src/modules/forms/TextAreaScreen.tsx:1]: Can we have the TextArea Sizes x TextArea Theme in order?  That way it is easier to check. @crstnmac

File: src/theme/defaultTheme/textArea.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
+export const textArea = {
+  wrapper: "relative w-full",
+  suffixWrapper:
+    "flex items-center justify-end pb-2 absolute inset-y-0 right-0 bg-transparent z-10",
+  size: {
+    sm: {
+      base: {
+        default: "h-[73px] p-1 text-cxs rounded-lg",
```
Comment (Karthik-B-06) [inline src/theme/defaultTheme/textArea.ts:8]: @crstnmac Can you check once if the padding for all different sizes is the same as the design file?

File: example/src/modules/forms/TextAreaScreen.tsx (hunk 1 above)
Comment (crstnmac) [reply example/src/modules/forms/TextAreaScreen.tsx:1]: Sorted the components

File: src/theme/defaultTheme/textArea.ts (hunk 2 above)
Comment (crstnmac) [reply src/theme/defaultTheme/textArea.ts]: fixed the padding and borderRadius according to the design

File: example/src/modules/forms/TextAreaScreen.tsx (hunk 1 above)
Comment (Karthik-B-06) [reply example/src/modules/forms/TextAreaScreen.tsx:1]: Great, Can you use a Radio component to switch between sizes and variants, and just one TextArea on the screen? That way it's neater.

Comment (crstnmac) [reply example/src/modules/forms/TextAreaScreen.tsx:1]: <image src="https://user-images.githubusercontent.com/20360870/201634352-968b1ae4-356d-47ef-8475-4de09a9fb9dd.png" height="800" width="350" />

File: src/theme/defaultTheme/textArea.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        base: "bg-gray-100 border-gray-400",
+        placeholder: "text-gray-500",
+        suffix: "text-gray-500",
+      },
+      invalid: {
+        base: "border-red-400",
+        suffix: "text-red-800",
+      },
+    },
+    subtle: {
```
Comment (Karthik-B-06) [inline src/theme/defaultTheme/textArea.ts:73]: `subtle` seems to be not working.

File: src/components/text-area/TextArea.tsx [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+} from "../../utils";
+import { createComponent } from "../../utils/createComponent";
+import { composeEventHandlers } from "../../utils/mergeRefs";
+import { Spinner } from "../spinner";
+import { createIcon, Icon, TextAreaSuffix } from "..";
+
+export type TextAreaVariants = "outline" | "subtle" | "underline" | "ghost";
+export type TextAreaSizes = "sm" | "md" | "lg" | "xl";
+
+export interface TextAreaProps extends TextInputProps {
```
Comment (Karthik-B-06) [inline src/components/text-area/TextArea.tsx:23]: I think we have not handled the `disabled` case.

Comment (crstnmac) [reply src/components/text-area/TextArea.tsx:23]: editable prop is there for the disabled case

File: src/theme/defaultTheme/textArea.ts (hunk 3 above)
Comment (crstnmac) [reply src/theme/defaultTheme/textArea.ts:73]: Fixed subtle styles for input & text-area

File: src/components/text-area/TextArea.tsx (hunk 4 above)
Comment (Karthik-B-06) [reply src/components/text-area/TextArea.tsx:23]: Oh, then add it as a controllable.

Comment (crstnmac) [reply src/components/text-area/TextArea.tsx:23]: <image src="https://user-images.githubusercontent.com/20360870/201653348-0b18c92d-166f-4ebc-a31b-5154873b3cd7.png" height="800" width="350" />

Comment (crstnmac) [reply src/components/text-area/TextArea.tsx:23]: Also added disabled state

File: src/components/text-area/TextArea.tsx [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+  styleAdapter,
+  useOnFocus,
+  useOnHover,
+} from "../../utils";
+import { createComponent } from "../../utils/createComponent";
+import { composeEventHandlers } from "../../utils/mergeRefs";
+import { Spinner } from "../spinner";
+import { createIcon, Icon, TextAreaSuffix } from "..";
+
+export type TextAreaVariants = "outline" | "subtle" | "underline" | "ghost";
```
Comment (Karthik-B-06) [inline src/components/text-area/TextArea.tsx:20]: We can remove `ghost` as its not present in the design system

Comment (crstnmac) [reply src/components/text-area/TextArea.tsx:20]: okay

**Comment by Karthik-B-06** [comment]
Add all A11y props for TextArea on the web. @crstnmac 
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/textbox_role

File: example/src/modules/forms/TextAreaScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  Switch,
+  TextArea,
+  TextAreaSizes,
+  TextAreaVariants,
+  useTheme,
+} from "@adaptui/react-native-tailwind";
+
+export const TextAreaScreen = () => {
+  const tailwind = useTheme();
+  const [variant, setVariant] = useState("subtle");
```
Comment (Karthik-B-06) [inline example/src/modules/forms/TextAreaScreen.tsx:17]: Use types here than using `as <type>` while sending it as props to the component.

File: example/src/modules/forms/TextAreaScreen.tsx [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+          <Radio value="xl" label="Extra Large" />
+        </RadioGroup>
+        <Box style={tailwind.style("flex-row justify-center")}>
+          <Box style={tailwind.style("flex-col items-center")}>
+            <Switch
+              label="Loading"
+              state={loading}
+              onStateChange={setLoading}
+            />
+            <Switch
```
Comment (Karthik-B-06) [inline example/src/modules/forms/TextAreaScreen.tsx:63]: The invalid switch seems to be disabled. 

https://user-images.githubusercontent.com/35562287/201837829-876908a8-a5d3-4522-9ba4-173267e69462.mov

File: src/components/text-area/TextArea.tsx [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+  state: DefaultTextAreaSpinnerProps,
+): JSX.Element => {
+  const { size, spinnerStroke } = state;
+  return <Spinner size={size !== "xl" ? "xs" : "md"} style={spinnerStroke} />;
+};
+
+const RNTextArea: React.FC<Partial<TextAreaProps>> = forwardRef<
+  typeof RNTextInput,
+  Partial<TextAreaProps>
+>((props, ref) => {
```
Comment (Karthik-B-06) [inline src/components/text-area/TextArea.tsx:80]: @crstnmac  The `ref` over here does not get assigned to the Input component, I think we need to do a 
`mergeRefs([inputRef, ref])` , 

```
const localRef = useRef();
const inputRef = mergeRefs([ref, localRef])

// Assign this to the input component
<Input
...
...
ref={inputRef}
/>
```

I think with this, we can make the `ref.blur()` work.

File: src/components/text-area/TextArea.tsx [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+          : textAreaTheme.variant[variant]?.disabled?.placeholder,
+      ),
+    );
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isFocussedWeb, isHovered, editable, isFocussedMobile]);
+
+  React.useEffect(() => {
+    // @ts-ignore
+
+    ref?.current?.setNativeProps?.({
```
Comment (Karthik-B-06) [inline src/components/text-area/TextArea.tsx:132]: We will have to change this ref over here to the mergedRef value. @crstnmac

File: src/components/text-area/TextArea.tsx [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    textInputWrapperProps = {},
+    prefix,
+    suffix,
+    invalid = false,
+    editable = true,
+    loading,
+    _suffixProps,
+    value,
+    style: textInputStyle,
+    accessibilityLabel,
```
Comment (Karthik-B-06) [inline src/components/text-area/TextArea.tsx:100]: Set a default a11y label value. @crstnmac

Comment (crstnmac) [reply src/components/text-area/TextArea.tsx]: Done

File: src/components/text-area/TextArea.tsx [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+            : {},
+          styleAdapter(textInputStyle),
+        ]}
+        placeholderTextColor={placeholderTextColor}
+        editable={editable}
+        {...restProps}
+        onFocus={composeEventHandlers(onFocus, handleOnFocus)}
+        onBlur={composeEventHandlers(onBlur, handleOnBlur)}
+        // A11y Props
+        accessibilityLabel={accessibilityLabel}
```
[the remark this answers is not in the document]
Comment (crstnmac) [reply src/components/text-area/TextArea.tsx:264]: Done

File: src/components/text-area/TextArea.tsx [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+  /**
+   * Suffix wrapper props extends Touchable Props
+   */
+  _suffixProps: TouchableProps;
+  /**
+   * A11y Label
+   */
+  accessibilityLabel: string;
+  /**
+   * A11y Described By
```
Comment (Karthik-B-06) [inline src/components/text-area/TextArea.tsx:65]: Better explanation of A11y props will be good.

File: src/components/text-area/TextArea.tsx (hunk 8 above)
Comment (Karthik-B-06) [reply src/components/text-area/TextArea.tsx:92]: Did it work?

File: src/components/text-area/TextArea.tsx [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+        placeholderTextColor={placeholderTextColor}
+        editable={editable}
+        {...restProps}
+        onFocus={composeEventHandlers(onFocus, handleOnFocus)}
+        onBlur={composeEventHandlers(onBlur, handleOnBlur)}
+        // A11y Props
+        accessibilityLabel={accessibilityLabel}
+        // A11y Props
+        defaultValue={defaultValue}
+        // @ts-ignore
```
Comment (Karthik-B-06) [inline src/components/text-area/TextArea.tsx:267]: Explain why we have used @ts-ignore. [Just Better commenting]

File: src/theme/defaultTheme/input.ts [hunk 14]
```diff
@@ -96,7 +96,7 @@ export const input = {
         suffix: "text-gray-700 ",
       },
       active: {
-        base: "bg-white-900",
```
Comment (Karthik-B-06) [inline src/theme/defaultTheme/input.ts:99]: Why was border-gray-500 removed ?

Comment (crstnmac) [reply src/theme/defaultTheme/input.ts:99]: It was not there in Input style so added it

**Comment by Karthik-B-06** [comment]
@crstnmac Add neat previews to the docs. Similar to other components

**Comment by crstnmac** [comment]
SCR-20221115-utg

File: docs/text-area.md [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+export default function App() {
+  return (
+      <TextArea placeholder="Type Something...." />
+  )
+}
+```
+
+## Table of Contents
+
+- [Textarea](#textarea)
```
Comment (Karthik-B-06) [inline docs/text-area.md:20]: I think Table of contents can just have the sections below it and not above.

File: docs/text-area.md [hunk 16]
```diff
@@ -0,0 +1,113 @@
+# Textarea
+
+Adapt UI provides textarea with three styles and six interaction
+states. Use this component to receive information from the user.
+
+![SCR-20221115-utg](https://user-images.githubusercontent.com/20360870/201976908-c51596e9-0ac6-4740-a2ba-660e6f9d8ca7.png)
```
Comment (Karthik-B-06) [inline docs/text-area.md:6]: A better preview, taken in a landscape, With the heading and the hamburger menu, would be nice to have.

Comment (crstnmac) [reply docs/text-area.md:6]: SCR-20221116-gpf

Comment (crstnmac) [reply docs/text-area.md:6]: SCR-20221116-gzi

File: docs/text-area.md [hunk 17]
```diff
@@ -0,0 +1,168 @@
+# Textarea
```
Comment (Karthik-B-06) [inline docs/text-area.md:1]: Can you please add all the previews as landscaped images?

File: docs/text-area.md [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+
+### Usage
+
+```js
+import { TextArea } from "@adaptui/react-native-tailwind"
+
+export default function App() {
+  return (
+    <>
+      <TextArea placeholder="Type something" />
```
Comment (Karthik-B-06) [inline docs/text-area.md:79]: Having just "Type something" everywhere seems boring. 
Try to have some real use cases in previews everywhere. @crstnmac

File: example/src/modules/forms/TextAreaScreen.tsx [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+              label="Invalid"
+              state={invalid}
+              onStateChange={setInvalid}
+            />
+          </Box>
+          <Box style={tailwind.style("flex-col items-center")}>
```
Comment (Karthik-B-06) [inline example/src/modules/forms/TextAreaScreen.tsx]: A nice idea here. 💡 Can you swap the positions to have a focus in button 1st? @crstnmac

**Comment by crstnmac** [comment]
simulator_screenshot_CD5180E6-6A0E-4F15-A23E-69E7E64065D8

File: docs/text-area.md (hunk 17 above)
Comment (crstnmac) [reply docs/text-area.md:1]: added

File: docs/text-area.md (hunk 18 above)
Comment (crstnmac) [reply docs/text-area.md:80]: ok

File: example/src/AppRoot.tsx [hunk 20]
```diff
@@ -103,19 +104,29 @@ const AppRoot = () => {
...
+        options={{ title: "Text Area" }}
+        name="TextAreaScreen"
+        component={TextAreaScreen}
+      />
+      <Drawer.Screen
+        options={{ title: "Tooltip" }}
+        name="TooltipScreen"
+        component={TooltipScreen}
+      />
+      <Drawer.Screen
```
[the remark this answers is not in the document]
Comment (crstnmac) [reply example/src/AppRoot.tsx]: resolved it

File: example/src/modules/forms/TextAreaScreen.tsx (hunk 19 above)
Comment (crstnmac) [reply example/src/modules/forms/TextAreaScreen.tsx]: ok

File: example/src/modules/forms/TextAreaScreen.tsx [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+        placeholder={"Type Something...."}
+        size={size}
+        variant={variant}
+        loading={loading}
+        invalid={invalid}
+        suffix={suffix}
+        editable={!disabled}
+        ref={ref}
+      />
+      <Box style={tailwind.style("absolute bottom-0")}>
```
Comment (Karthik-B-06) [inline example/src/modules/forms/TextAreaScreen.tsx]: Can you please Make the radio group items centered?  @crstnmac

File: example/src/modules/forms/TextAreaScreen.tsx [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+          onChange={value => setSize(value as TextAreaSizes)}
+          orientation="horizontal"
+        >
+          <Radio value="sm" label="Small" />
+          <Radio value="md" label="Medium" />
+          <Radio value="lg" label="Large" />
+          <Radio value="xl" label="Extra Large" />
+        </RadioGroup>
+        <Box style={tailwind.style("flex-row justify-center")}>
+          <Box style={tailwind.style("flex-col items-center")}>
```
Comment (Karthik-B-06) [inline example/src/modules/forms/TextAreaScreen.tsx]: Also, can you please add spacing between switch components? @crstnmac some breathing space. Its so packed here now.

Comment (crstnmac) [reply example/src/modules/forms/TextAreaScreen.tsx]: Updated code

**Comment by crstnmac** [comment]
Final look
<image src="https://user-images.githubusercontent.com/20360870/202177838-dccd168e-f8be-4523-8c13-f34fbf3300e1.png" width="350" height="800" />

File: docs/text-area.md (hunk 17 above)
Comment (Karthik-B-06) [reply docs/text-area.md:1]: There seems to be a spelling mistake in the placeholder. @crstnmac

File: docs/text-area.md [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+
+## Variant
+
+Adapt UI provides four different input field components styles: `outline`,
+`subtle` and `underline`.
+
+You can use these various styled input field components based on the necessity
+and style you wish to provide to your design.
+
+![simulator_screenshot_93D9FFAA-AC38-4CA6-94F0-E7A861FF5187](https://user-images.githubusercontent.com/20360870/202174447-930c4d36-5f12-448e-bd19-44f9d75d19b8.png)
```
Comment (Karthik-B-06) [inline docs/text-area.md:64]: The order above mentioned above and the preview do not match.

File: docs/text-area.md [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+</details>
+
+## Size
+
+There are four different sizes for input in Adapt UI: `sm`, `md`, `lg` & `xl`
+
+Based on the hierarchy, you can switch between different sizes.
+
+
+![simulator_screenshot_BCCD5B62-43DE-4E94-934A-97A6D2C9956F](https://user-images.githubusercontent.com/20360870/202175990-974369e3-4880-41d2-857d-fc55aeb1700e.png)
```
Comment (Karthik-B-06) [inline docs/text-area.md:98]: Can you try to squeeze the components a little bit to place them correctly on the screen?

File: docs/text-area.md (hunk 23 above)
Comment (crstnmac) [reply docs/text-area.md:64]: fixed it

**Comment by Karthik-B-06** [comment]
@crstnmac In one of the previews, there is a spelling mistake in the placeholder text (yourself*).  

Screenshot 2022-11-18 at 10 08 57 AM

**Review by Karthik-B-06 — APPROVED** [review]
LGTM 🎉

**Pull request merged**